### PR TITLE
Add FeatureSelection tool

### PR DESCRIPTION
**Pull request by chrismayer**
This adds a FeatureSelection tool as described in #245.

245-feature-select

### CODE REVIEW

**Comment by chrismayer** [comment]
Some remarks: 

- Due to clustering of the vector features it sometimes looks like just one feature was clicked but in the grid two or more appear. 
- Would it be useful to inform the user if no feature was hit by the map click? Currently anything is done. Maybe a little feedback would increase usability?

**Comment by simonseyock** [comment]
As long as your interaction is active you can disable the default behaviour by calling
```
olMap.set('enableDefaultClick', false);
```
and enable it again with setting it to true.

File: app/controller/button/FeatureSelectionButton.js [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+        var me = this;
+
+        if (!me.modeSelector) {
+            me.modeSelector = Ext.create('Ext.button.Split', {
+                text: 'Add to selection',
+                hidden: true,
+                menu: new Ext.menu.Menu({
+                    items: [
+                        {
+                            text: 'Add to selection',
```
Comment (geographika) [inline app/controller/button/FeatureSelectionButton.js:100]: Change to "Add to Selection"

File: app/controller/button/FeatureSelectionButton.js [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+
+    /**
+     * Adds the mode selector UI to the toolbar (if not existing) and shows it.
+     */
+    addModeSelectorUi: function () {
+        var me = this;
+
+        if (!me.modeSelector) {
+            me.modeSelector = Ext.create('Ext.button.Split', {
+                text: 'Add to selection',
```
Comment (geographika) [inline app/controller/button/FeatureSelectionButton.js:95]: Change to "Add to Selection"

File: app/controller/button/FeatureSelectionButton.js [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+                    items: [
+                        {
+                            text: 'Add to selection',
+                            handler: function (menu) {
+                                me.filterMode = 'ADD_TO_SELECTION';
+                                me.modeSelector.setText(menu.text);
+                            }
+                        },
+                        {
+                            text: 'New selection',
```
Comment (geographika) [inline app/controller/button/FeatureSelectionButton.js:107]: Change to "New Selection"

**Review by geographika — COMMENTED** [review]
@chrismayer - playing with this now! Looks excellent. 
I'll need some time to try and test integration in the larger application.

A couple of initial comments. 

I might change the `selModel` on the grid to `MULTI` as then you can easily highlight all features in the grid on the map. Any foreseeable problems with this?

I think the clustering issue is fine.

In the old system when a feature was not selected I had an Ext animation flash a message in a status bar at the bottom of the screen. Maybe if an event was fired this functionality could be added in a future release.

File: app/controller/button/FeatureSelectionButton.js (hunk 3 above)
Comment (chrismayer) [reply app/controller/button/FeatureSelectionButton.js]: Done in 4f03a22

File: app/controller/button/FeatureSelectionButton.js (hunk 2 above)
Comment (chrismayer) [reply app/controller/button/FeatureSelectionButton.js]: Done in 4f03a22

File: app/controller/button/FeatureSelectionButton.js (hunk 1 above)
Comment (chrismayer) [reply app/controller/button/FeatureSelectionButton.js]: Done in 4f03a22

File: app/controller/button/FeatureSelectionButton.js [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+
+        // clear FIDs to filter when we have a new selection
+        if (me.filterMode === 'NEW_SELECTION') {
+            me.fidsToFilter = [];
+        }
+
+        // collect all IDs of clicked features
+        me.map.forEachFeatureAtPixel(evt.pixel, function(feature, layer) {
+            // add check for correct layer
+            if (layer.id === view.queryLayer.id) {
```
[the remark this answers is not in the document]
Comment (chrismayer) [reply app/controller/button/FeatureSelectionButton.js]: Done in 4f03a22

**Comment by chrismayer** [comment]
I addressed the review comments of @geographika.

Also removed the conflict with the GetFeatureInfo tool and added some behaviour if no feature was hit by clicking on the  map:

  - The view fires an event `'cmv-no-feature-selected'` 
  - By setting the views config option `showNoSelectionMessage` to `true` a little `Toast` is shown in the lower-right corner of the app with a message

**Comment by chrismayer** [comment]
> I might change the selModel on the grid to MULTI as then you can easily highlight all features in the grid on the map. Any foreseeable problems with this?

I don't think so because this does not interact with the selection model at all.

**Comment by geographika** [comment]
@chrismayer - I may make a follow-up pull request to allow a grid to have just the "Select by Polygon" or both select tools - but please merge this as-is - thanks.

**Pull request merged**